### PR TITLE
Add 2D navigation mesh baking

### LINKED ISSUE #5508: Add 2D navigation mesh simplifcation to 4.0 TileMap
### Describe the project you are working on

Experimenting with the TileMap and Navigation features
Planning to use them in procedural map generation (not necessarily randomised)

### Describe the problem or limitation you are having in your project

When pathing across a TileMap defined NavigationMap the pathing picks paths that only cross between adjacent or diagonal tiles
This leads to inefficient paths and effectively reduces pathing to simple 8-directional, rather than 360 degree movement, as tile size trends towards 1px

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Access to a TileMap's NavigationMesh and built in "simplify navigation mesh" function would solve this issue even at it's most primitive, and provide other useful functionality.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

It is currently hypothetically possible to iterate over each tile in a TileMap (or navigation Mesh): look for adjacent navigation polygons, check they share vertices and manually merge them to slowly form a new NavigationMesh.
This is both quite cumbersome and would become painfully slow in GDscript with larger TileMaps

Having at least full tiles flag themselves to be merged into adjacent meshes would be very useful.
If TileMap received "set_tiles_rect" and/or "set_tiles_fill" this could be further simplified by creating the NavigationPolygon using just the edge tiles.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not to my knowledge.
Even in the case of a smaller function in terms of lines of code it would be iterating over thousands of nodes in a cumbersome way to handle in GDscript without special cases.

A general "simplify mesh" function could see a lot more general use.

### Is there a reason why this should be core and not an add-on in the asset library?

This would make TileMap a far more powerful tool to handle graphics, static body physics physics, and navigation.
Having TileMap handle these features without needing custom code would open up the features for simpler use by artists and designers without experience in and/or are intimidated by coding.

### LINKED ISSUE #66283: `Condition "least_cost_id == -1" is true.` Error using `NavigationAgent2D` on `TileMap` for `autotile`
### Godot version

v3.5.stable.official [991bb6a]

### System information

Windows 10

### Issue description

When I use `NavigationAgent2D` on `autotile`, it works weird and keeps emitting a lot of error:
```
E 0:00:01.973   get_path: Condition "least_cost_id == -1" is true. Breaking.
  <C++ Source>  modules/navigation/nav_map.cpp:263 @ get_path()
  <Stack Trace> Player.gd:12 @ _physics_process()
```
bug
nav
Sometimes the `Player` moves and sometimes it doesn't. I can't describe it simply, check minimal reproduction project for details.

### Steps to reproduce

Run MRP. It has 3 Scenes, with the same code, different type of map.
- `World_bug`: use autotile, works weird and emits error.
- `World_tile`: use single tile, works fine.
- `World_poly`: use navigation polygon, works fine.

### Minimal reproduction project

[bug.zip](https://github.com/godotengine/godot/files/9630582/bug.zip)

### LINKED ISSUE #70823: NavigationPolygon make_polygons_from_outlines fails on specific inner outlines sizes combinations
### Godot version

v4.0.beta10.official [d0398f62f]

### System information

Windows 10

### Issue description

image

I'm making navigation region and I want to place a polygon inside a hole inside a polygon, like here ^
I expect to use big counterclockwise outline first, then add smaller clockwise outline, and finally add smallest counterclockwise outline. I create three outlines and everyting works fine.
```
	var poly1 = PackedVector2Array([Vector2(500, -50), Vector2(500, 500), Vector2(-50, 500), Vector2(-50, -50)])
	var poly2 = PackedVector2Array([Vector2(101, 101), Vector2(101, -5), Vector2(-5, -5), Vector2(-5, 101)])
	var poly3 = PackedVector2Array([Vector2(59, 59), Vector2(37, 59), Vector2(37, 37), Vector2(59, 37)])
```
And these outlines works fine if order is Counterclockwise(poly1)->Clockwise(poly2)->Clockwise(poly3) or if Counterclockwise(poly1)->Clockwise(poly2)->Counterclockwise(poly3)

But then I try to make another navigation region with bigger outer outline (poly1 array), wierd things happen.
I use 
```
	var poly1 = PackedVector2Array([Vector2(600, -50), Vector2(600, 600), Vector2(-50, 600), Vector2(-50, -50)])
	var poly2 = PackedVector2Array([Vector2(101, 101), Vector2(101, -5), Vector2(-5, -5), Vector2(-5, 101)])
	var poly3 = PackedVector2Array([Vector2(59, 59), Vector2(37, 59), Vector2(37, 37), Vector2(59, 37)])
```
If order is Counterclockwise(poly1)->Clockwise(poly2)->Clockwise(poly3) everything is fine. If order is Counterclockwise(poly1)->Clockwise(poly2)->Counterclockwise(poly3) i get an error
`NavigationPolygon outlines can not overlap vertices or edges inside same outline or with other outlines or have any intersections.
`

### Steps to reproduce

Open minimal reproduction project included, turn on "Visible navigation" in "Debug" menu
Press "500x500" button
Run agaun
Press "600x600" button

### Minimal reproduction project

[testtileminim.zip](https://github.com/godotengine/godot/files/10329632/testtileminim.zip)

### PR DESCRIPTION
**Pull request by smix8**
Adds 2D navigation mesh baking for `NavigationRegion2D` and `NavigationServer2D`.

Slight changed behavior on the baking process of the NavigationPolygon, see info further below.

navmesh2d_baking

Supersedes https://github.com/godotengine/godot/pull/70724.

### Fixed Issues
- Fixes https://github.com/godotengine/godot/issues/66283 (through bake and merge step).
- Fixes https://github.com/godotengine/godot/issues/70823 (through baking replacing the old function).

### Related older Issues
- https://github.com/godotengine/godot/issues/72518.
- https://github.com/godotengine/godot/issues/57967.
- https://github.com/godotengine/godot/issues/69743.

### Implemented proposals
- Closes https://github.com/godotengine/godot-proposals/issues/5508 Add 2D navigation mesh simplifcation to 4.0 TileMap (through NavigationRegion2D baking for TileMap)

### New 2D navigation mesh baking

The navigation mesh baking for 2D is accessible through `NavigationRegion2D`.

The NavigationRegion2D has new buttons in the Editor for baking or clearing the current NavigationPolygon.
The NavigationRegion2D also has a new function `bake_navigation_polygon()` to bake the navigation polygon with scripts.

navigationregion2dbakemenu

The 2D navigation mesh baking workflow is in general kept very similar to 3D but there are few key differences.

2D requires at least 1 bounding outline that defines the outer limits of the parsed area and baked surface.
Everything inside those bounds is by default traversable surface and all parsed objects are considered obstructions.

navregion2d

Another, more advanced way to bake navigation meshes for 2D is available for scripts when using the `NavigationServer2D` singleton directly. The `NavigationServer2D` has new, dedicated 2D functions for parsing and baking 2D navigation meshes.

- `parse_source_geometry_data()` can be used to only parse source geometry to a reuseable and serializable `NavigationMeshSourceGeometryData2D` resource without the baking step (see description below).
- `bake_from_source_geometry_data()` can be used to bake a NavigationPolygon from already parsed data e.g. to avoid runtime performance issues with (redundant) parsing.
- `bake_from_source_geometry_data_async()` is the same but bakes the NavigationPolygon deferred with threads, not blocking the main thread.

The source geometry parsing is implemented with dedicated parsers for each supported node type. Currently the following node parsers are added:
- StaticBody2D and all CollisionShape2D
- Polygon2D
- TileMap (Layer0 only for now)
- MeshInstance2D (2D meshes only)
- MultiMesh2D (2D meshes only)

Note that only 2D meshes are parsed and all 3D meshes are ignored. Meshes are considered 2D when the mesh format `ARRAY_FLAG_USE_2D_VERTICES` flag is set.

Note that all rounded default shapes have been hardcoded to far low amount of edges compared to their visuals.

- Circle shapes use 12 edges
- Capsule shapes use 14 edges

Too many edges not only drag the bake performance down but also result in polygon "star" artifacts in the navmesh that cause pathfinding issues.

In general rounded or other overdetailed objects are not recommended to be used as source geometry for navigation mesh baking and should all be replaced by polygons with a more simplified shape. Also prefer physics collision shapes over visual polygons.

### TileMap navigation mesh baking

The TileMap is a special case for 2D navigation mesh baking as it is a node that combines internally both traversable surfaces as well as obstructions. Baking a TileMap has certain limitations due to the current TileMap design.

- Only the first TileMap Layer 0 is considered in the navigation mesh baking.
- TileMap Layer 0 cells with a navigation polygon count as traversable.
- TileMap Layer 0 cells with a collision polygon count as obstruction.

A NavigationRegion2D can bake a navigation mesh for the entire TileMap by merging the navigation polygons and collision polygons from all used TileMap cells on the first Tilemap layer.

217048075-9cd9c4c5-5dfa-4c76-bf86-ea64331caeca

This full TileMap cell merge is required as it is the only reliable way to get working navigation meshes that are fitted for an agent size. It yields the best possible result for both navigation performance and quality by removing all those little TileMap cell edge seams that are responsible for a lot of pathfinding and performance problems with the TileMap build-in navigation.

### NavigationMeshSourceGeometryData2D

NavigationMeshSourceGeometryData is the resulting data of a parsing operation done with the NavigationServer used for navigation mesh baking.

The advantage of having this data available in a resources is that it can be stored and loaded from disk or resued. This helps to avoid runtime performance issues with parsing operations on larger scenes. NavigationMeshSourceGeometryData makes it possibility to split the source geometry parsing process from the navigation mesh baking process. It also makes it possible to reuse the same source data to bake multiple meshes with different parameters, e.g. use one source geometry to bake navigation meshes for multiple different agent sizes.

### Compatibility

The NavigationPolygon has some changed behavior now.

- The Editor drawn outlines now define the surfaces that should be baked to a 2D navigation mesh.
- Both traversable outlines and obstruction outlines are merged before the baking step, e.g. they can now have cross edges without errors. This also means that outlines "nested" inside other outlines will no longer create holes by default as this was just a constant source of user bugs. If you want reliable holes add an obstruction.
- Dragging outlines in the Editor will not auto-bake polygons all the time, hit the bake button when you are done editing. The old auto-bake was really laggy when editing more complex polygons.

The `NavigationPolygons.make_polygons_from_outlines()` function is deprecated. It still works like before with all known bugs and issues. If possible consider upgrading old projects to the new baking with the NavigationServer.

Existing NavigationPolygons that were already baked with `make_polygons_from_outlines()` are compatible.

Existing outlines are also compatible but if baked with the new bake the result might be interpreted differently when nested "hole" outlines were used. Consider using obstacle outlines to cut holes in the traversable polygons instead of nesting outlines into other outlines which has always a chance to flip the polygon in unintended ways.

Old NavigationPolygon that have at least one outline will bake just fine with the new bake but since the NavigationPolygon by default has an agent radius of 10 pixels this will shrink new polygons. If you don't want this set the agent radius to zero to create polygons similar to the old polygons.

### Internals

- Adds a NavigationServer2D dummy same as the NavigationServer3D dummy.
- Makes the NavigationServer2D pure virtual with a GodotNavigationServer2D override.
- NavigationServer `sync()` point to dispatch async baked navmeshes now happens each iteration just before the physics loop. The old had it as part of the physics process loop which made it unreliable and frustrating to use.

### Thirdparty
This pr includes thirdparty Clipper2 library as a requirement for the polygon operations without touching any of the existing Clipper1 uses in the engine. 

I couldn't get any good results with the old Clipper1 as it broke far to often with edge cases or very complex polygons and was in general also way too slow for runtime navmesh baking so upgrading was imo mandatory.

Clipper1 is currently used across the engine e.g. the Geometry2D class and some Sprite ops. I added Clipper2 to the core thirdparty so it is available for the rest of the engine so all those old parts can be upgraded over time. Since I never really use those 2D functions I don't want to touch them myself.

### CODE REVIEW

**Comment by vpellen** [comment]
Been waiting for this, glad you managed to pull past the various breaking changes!

Question: Does this do a bunch of polygon ops or does it do the same thing that the 3D version does and do a bunch of point casts? I'd assume it's the former based on circle edges and the fact that you're using a polygon clipping library, but if that's the case, the "cell size" parameter confuses me, as I can't imagine it'd be relevant.

Also, regarding circle edge density: Have you considered making it an adjustable factor of their radius?
Also also, is the agent radius just an inverse extrude on the final mesh? Would I be correct in assuming that was what was causing some of the miserable clipping issues that warranted an external lib upgrade?

**Comment by smix8** [comment]
> Question: Does this do a bunch of polygon ops ...

It uses Clipper2 polygon ops like Union and Difference under the hood.

> ...  the "cell size" parameter confuses me, as I can't imagine it'd be relevant.

The `NavigationPolygon.cell_size` property is primarily for the navmesh merge on the navigation map on the NavigationServer. The navigation map also has a cell_size used to rasterize the navmesh edges by edgekey to merge them. If the NavigationPolygon cell_size and navigation map cell_size are mismatched merge errors can occur.

Previously the cell_size was also used to upscale the arrays as is required for the rasterization of the Clipper1 lib to avoid bugs but this slowed things down considerably. After switching to Clipper2 I found no reason to do this anymore, instead found one more reason why Clipper1 is so obsolete.

> ... is the agent radius just an inverse extrude on the final mesh?

The `agent_radius` is the deflating amount that the final navmesh polygon is shrunk when baking it.

**Comment by KoB-Kirito** [comment]
> Only the first TileMap Layer 0 is considered in the navigation mesh baking

Would it be very hard to consider all layers?  
All layers can contain collision. You would have to create an extra layer and layout all collisions there again just to bake the map, which is immense work in some scenarios.

**Comment by smix8** [comment]
> Would it be very hard to consider all layers?

It is primarily a user interface issue as the parsing process does not care what it parses but it also can not make random assumptions how a TileMap user intends to use and mix its TileMap layers. This is the entire reason this limitation exists so poke the TileMap devs if you want this limitation lifted at some point.

**Comment by ghmart** [comment]
I've set up some crazy experiment: using NavigationPolygon for GridMap. This is a project, that allows adding / destroying cells with navigation at runtime. I'm not using standard baking, rolled my own algorithm, that utilizes "make_polygons_from_outlines" at some stage. Sometimes it fails to detect holes / islands, say 10% of time.

I tried this project with your version (with some changes - multiply cells' coordinates by 100, for example, because it can't create navigation mesh with floating-point coordinates, maybe because this version uses integer Point64 in Clipper2). Still it fails more often than with "make_polygons_from_outlines", maybe around 30% of time or even more. Sometimes it even can't handle similar cases.

navigation

So now, I gave up to find the cause of error. Maybe it's in Clipper2 library, or in my algorithm (maybe need to use only positive coordinates), or in new version of navigation mesh creation.

Anyway, would be great to have new mesh baking version, alongside Clipper2 in Godot!

File: doc/classes/NavigationPolygon.xml [hunk 1]
```diff
@@ -1,44 +1,44 @@
...
-		var polygon = new NavigationPolygon();
-		var outline = new Vector2[] { new Vector2(0, 0), new Vector2(0, 50), new Vector2(50, 50), new Vector2(50, 0) };
-		polygon.AddOutline(outline);
-		polygon.MakePolygonsFromOutlines();
-		GetNode&lt;NavigationRegion2D&gt;("NavigationRegion2D").NavigationPolygon = polygon;
+		var new_navigation_mesh = new NavigationPolygon();
+		var bounding_outline = new Vector2[] { new Vector2(0, 0), new Vector2(0, 50), new Vector2(50, 50), new Vector2(50, 0) };
+		new_navigation_mesh.AddOutline(bounding_outline);
+		NavigationServer2D.BakeFromSourceGeometryData(new_navigation_mesh, new NavigationMeshSourceGeometryData2D());
+		GetNode&lt;NavigationRegion2D&gt;("NavigationRegion2D").NavigationPolygon = new_navigation_mesh;
```
Comment (AThousandShips) [inline doc/classes/NavigationPolygon.xml:22]:
```suggestion
		var newNavigationMesh = new NavigationPolygon();
		var boundingOutline = new Vector2[] { new Vector2(0, 0), new Vector2(0, 50), new Vector2(50, 50), new Vector2(50, 0) };
		newNavigationMesh.AddOutline(boundingOutline);
		NavigationServer2D.BakeFromSourceGeometryData(newNavigationMesh, new NavigationMeshSourceGeometryData2D());
		GetNode&lt;NavigationRegion2D&gt;("NavigationRegion2D").NavigationPolygon = newNavigationMesh;
```

File: doc/classes/NavigationPolygon.xml [hunk 2]
```diff
@@ -1,44 +1,44 @@
...
-		polygon.Vertices = vertices;
-		var indices = new int[] { 0, 1, 2, 3 };
-		polygon.AddPolygon(indices);
-		GetNode&lt;NavigationRegion2D&gt;("NavigationRegion2D").NavigationPolygon = polygon;
+		var new_navigation_mesh = new NavigationPolygon();
+		var new_vertices = new Vector2[] { new Vector2(0, 0), new Vector2(0, 50), new Vector2(50, 50), new Vector2(50, 0) };
+		new_navigation_mesh.Vertices = new_vertices;
+		var new_polygon_indices = new int[] { 0, 1, 2, 3 };
+		new_navigation_mesh.AddPolygon(new_polygon_indices);
+		GetNode&lt;NavigationRegion2D&gt;("NavigationRegion2D").NavigationPolygon = new_navigation_mesh;
```
Comment (AThousandShips) [inline doc/classes/NavigationPolygon.xml:41]:
```suggestion
		var newNavigationMesh = new NavigationPolygon();
		var newVertices = new Vector2[] { new Vector2(0, 0), new Vector2(0, 50), new Vector2(50, 50), new Vector2(50, 0) };
		newNavigationMesh.Vertices = newVertices;
		var newPolygonIndices = new int[] { 0, 1, 2, 3 };
		newNavigationMesh.AddPolygon(newPolygonIndices);
		GetNode&lt;NavigationRegion2D&gt;("NavigationRegion2D").NavigationPolygon = newNavigationMesh;
```

**Comment by ghmart** [comment]
Holes didn't recognized when manually adding points in clockwise order.

manual

**Comment by RafaelVidaurre** [comment]
I absolutely need this, gogogogog

**Comment by Calinou** [comment]
@RafaelVidaurre Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by JustMog** [comment]
root node children doesn't seem to work for me, also it tripped me up that the transform of the navregion isn't taken into account.
heres a repro project: [baketest.zip](https://github.com/godotengine/godot/files/12632855/baketest.zip)

**Comment by smix8** [comment]
Changes:
- Fixed various transform related bugs for some parsed node types.
- Limited the NavigationPolygon `cell_size` Editor slider range to full pixels.
It is still a float for technical reasons but the entire baking process happens on an integer grid so all your decimal fractional parts are ignored. You can still set floats with a script in case you have a custom navigation map cell_size that uses small float values but the baking will not support such small scaling and if you ignore this you might run into precision and navigation map sync errors.
- Reduced the amount of code indentations in the parse functions to make them more readable on small screens.

@JustMog 

> root node children doesn't seem to work for me

Since you added the "nav" group to both the NavigationPolygon and your main root Node2D in the scene I think  what you wanted to use instead was the `SOURCE_GEOMETRY_GROUPS_WITH_CHILDREN` "Group with Children" mode. This mode parses all nodes with this group_name that are inside the SceneTree and also parses recursive all the children of those nodes.

`SOURCE_GEOMETRY_ROOT_NODE_CHILDREN` "Root Node Children" does recursive parse the child nodes of the set `root_node`. This is **NOT** the root node of the SceneTree, it is whatever `root_node` is set with the parsing function and in case of baking with a `NavigationRegion2D.bake_navigation_polygon()` this node automatically sets itself as the root_node. In the repro project the NavigationRegion2D has no children in that scene so it only bakes the surface of your defined outline in that mode.

> it tripped me up that the transform of the navregion isn't taken into account

Some node types had wrong xform calculations but should be fixed with the last update, please retry.

**Comment by JustMog** [comment]
> This is **NOT** the root node of the SceneTree

ah, my mistake.
> Some node types had wrong xform calculations but should be fixed with the last update, please retry.

works now! thanks!

**Comment by smix8** [comment]
Added back everyone's favorite semi-broken function `NavigationPolygon.make_polygons_from_outlines()`.

The function is still deprecated and as broken as it was. It is only added back for compatibility for those projects that can not do without the old bake behavior for some reason. Everyone else enjoy the more robust NavigationServer baking.

**Comment by ghmart** [comment]
> Added back everyone's favorite semi-broken function `NavigationPolygon.make_polygons_from_outlines()`.
> 
> The function is still deprecated and as broken as it was. It is only added back for compatibility for those projects that can not do without the old bake behavior for some reason. Everyone else enjoy the more robust NavigationServer baking.

Not sure how broken was "make_polygons_from_outlines", but in my case it detects holes / islands automatically. I'd prefer to use new approach, but then I need to detect type of outlines myself like this:
```
var inside: int
for i in range(0, outlines.size()):
	inside = 0
	for j in range(0, outlines.size() - 1):
		if i == j:
			continue
		if Geometry2D.is_point_in_polygon(outlines[i][0], outlines[j]):
			inside += 1
	if inside % 2 == 0:
		if not Geometry2D.is_polygon_clockwise(outlines[i]):
			outlines[i].reverse()
	else:
		if Geometry2D.is_polygon_clockwise(outlines[i]):
			outlines[i].reverse()
```

**Comment by smix8** [comment]
> Not sure how broken was "make_polygons_from_outlines", but in my case it detects holes / islands automatically.

For starters as soon as you created overlap everything would break with the old bake. Also had many edge cases where even if the user placed everything correct it would still break. As you say it detects it only in many cases, this just not enough, you can't do a project with e.g. user placed objects and runtime rebake when there is a high chance that suddenly the entire navigation mesh is gone.

This is also why the new bake now merges everything first. This allows users to partially overlap their polygons without issues were the old bake would already break. Just do not fully place outlines inside each other which would again trigger the hole calculation.

That the holes sometimes work even with the new bake is more an afterthought because the underlying Clipper2 tries to figure out wtf the user tried to do here with so many outlines stacked inside each other. The result is still based on luck, because a single wrong placement is enough and the entire navigation mesh flips to the wrong side and interprets a large junk of the traversable polygon surface as a hole.

Holes should preferably added now by placing obstacle polygons, not by drawing inner outlines into other outlines. Because we merge the traversable polygons first, then merge the obstacles second, then get the diff between those two it is the far safer way to bake without stuff easily exploding in your face because a single vertex was placed on the edge of another polygon.

**Comment by ghmart** [comment]
Thank you for such detailed response. I agree with everything you said and will try using obstacles.

Main goal in my case was to speed up navigation mesh baking for GridMap. Most performance-critical part is in building outlines. Example in documentation ("Navmesh for 3D GridMaps") is unsuitable for real-time creation / destroying of cells with navigation. 

Maybe new baking system and new ability to use threads will change things.

File: thirdparty/README.md [hunk 3]
```diff
@@ -75,6 +75,16 @@ Files extracted from upstream source:
...
+## clipper2
+
+- Upstream: https://github.com/AngusJohnson/Clipper2
+- Version: 1.2.2  (756c5079aacab5837e812a143c59dc48a09f22e7, 2023)
+- License: Boost Software License 1.0
+
+Files extracted from upstream source:
+
+- `CPP/Clipper2Lib` folder
+- `LICENSE`
```
Comment (akien-mga) [inline thirdparty/README.md:87]: I don't see the license?

Comment (akien-mga) [reply thirdparty/README.md:87]: Also needs to be documented like this in `COPYRIGHT.txt`:
```diff
diff --git a/COPYRIGHT.txt b/COPYRIGHT.txt
index 3200eafa00..ce9be3dd5e 100644
--- a/COPYRIGHT.txt
+++ b/COPYRIGHT.txt
@@ -171,6 +171,11 @@ Comment: CA certificates
 Copyright: Mozilla Contributors
 License: MPL-2.0
 
+Files: ./thirdparty/clipper2/
+Comment: Clipper2
+Copyright: 2010-2013, Angus Johnson
+License: BSL-1.0
+
 Files: ./thirdparty/cvtt/
 Comment: Convection Texture Tools Stand-Alone Kernels
 Copyright: 2018, Eric Lasota
```

Comment (smix8) [reply thirdparty/README.md:103]: Seems I evaporated that file on a rebase by accident.

File: doc/classes/NavigationMeshSourceGeometryData2D.xml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		<method name="get_traversable_outlines" qualifiers="const">
+			<return type="PackedVector2Array[]" />
+			<description>
+				Returns all the traversable area outlines arrays.
+			</description>
+		</method>
+		<method name="has_data">
+			<return type="bool" />
+			<description>
+				Returns [b]true[/b] when parsed source geometry data exists.
```
Comment (akien-mga) [inline doc/classes/NavigationMeshSourceGeometryData2D.xml:47]:
```suggestion
				Returns [code]true[/code] when parsed source geometry data exists.
```

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by wayjack** [comment]
Doesn't seem like enabling/disabling NavigationRegion2Ds at runtime works anymore after this.

**Comment by YuriSizov** [comment]
@wayjack Please open an issue report with a reproduction project.

**Pull request merged**